### ISSUE TITLE
Volkov Commander incompatibility with SvarCOM

### ISSUE THREAD

**Issue by mateuszviste** [SvarDOS/bugz#41]
This is a placeholder for a ticket that has been once created by bttr at OSDN:
https://osdn.net/projects/svardos/ticket/45468

Unfortunately this page is not available for the moment and I am fuzzy on the details. To be clarified later.

**Comment by mateuszviste** [comment]
On 2022-08-27, bttr wrote:
> Hi Mateusz,
> 
> I think, I found an incompatibility between SvarCOM and Volkov Commander.
> 
> When returning to VC from a program I started from VC 4.05, e.g., my build of FED editor 2.24 or FreeDOS EDIT from the FreeDOS 1.3 release, I see some garbage on the screen.
> 
> This doesn't happen with FreeCOM 0.85a, also from FD 1.3.
> 
> I tested with SvarCOM 2022.3 and 2022.4. Also tried the SvarDOS default kernel in contrast to the FD 1.3 kernel 2043 (May 13 2021).
> 
> With VC 4.99.08 alpha things are even getting more worse: After quitting the "guest" program, I'm not in VC again, but at the DOS prompt.
> 
> When JEMM and other stuff is loaded, I often get a memory exception 06. Sometimes the machine hangs w/o or after the exception. I guess, either some register or memory region gets corrupted.
> 
> In VCBUG.7z you can find a floppy image with VC 4.05 and 4.99.08 alpha, FD kernel 2043, SvarCOM 2022.4, FED, FD-EDIT to play with.
> 
> Cheers,
> Robert

The original screenshots are unfortunately lost due to OSDN death, but the issue was relatively easy to reproduce. I attach the VCBUG.7z archive that Robert provided back then, as I had saved it (just repacked it into ZIP otherwise github does not accept it).

[VCBUG.zip](https://github.com/SvarDOS/bugz/files/13419208/VCBUG.zip)

**Comment by mateuszviste** [comment]
The same day (2022-08-27) I had responded:

> Hi Robert, thanks for the detailed trouble description. I have booted your image under VBox and I can confirm that: - with VC499, quitting the guest application returns directly to SvarCOM instead of spawning back a VC instance - with VC405, running EDIT, leaving it, and then quitting VC leaves a 3 weird characters on the console (similar to " = Ef")
> 
> When booting with FreeCOM as the command shell, none of these issue appear (ie. everything works as expected).
> 
> I didn't have any Exception 06, but that's because I didn't try loading JemmEx. Exception 06 means "invalid opcode" so it usually hints that either the codeflow jumped into some area of memory that wasn't supposed to be executed (like an invalid pointer set in an INT handler) or that part of the in-memory code has been corrupted by something.
> 
> From the top of my head I can only guess this might be related to how SvarCOM swaps in and out itself. VC certainly performs a similar task, and maybe they both end up screwing each other. FreeCOM performs no such swapping, which could explain why it works more reliably in this case.
> 
> I will investigate this in coming days.

(I was a bit optimistic about the ETA)

This might be related to <https://github.com/SvarDOS/bugz/issues/38> btw.

**Comment by mateuszviste** [comment]
The issue #38 ended up being unrelated to the VC problem, but ECM provided some very interesting hints on how one could try to track down the VC problem.

_Originally posted by @ecm-pushbx in https://github.com/SvarDOS/bugz/issues/38#issuecomment-1823107113_

> I may get around to debugging this later. My approach would be to run FreeCOM+VC+EDIT under the debugger (lDebug resident in `tsr` mode, or installed before FreeCOM), intercept some int 21h calls (`bp new ptr ri21p when value ah in 4B, 4C, 4D` plus `install indos`), and put breakpoints on where 21.4B calls should return (examine using `u ptr [ss:sp]` when an int 21h breakpoint has returned control to the debugger). Then you can examine the machine state using `dm` or `rvp` or `d ptr [psp:34] l word [psp:32]`. Next retry with SvarCOM instead and figure out what differs, other than the segment addresses.

**Comment by bttrx** [comment]
The missing screenshots:

Garbage after `A:\FED>` and `A:\EDIT>`:

![image](https://github.com/SvarDOS/bugz/assets/15858666/ccfb7698-f950-41fe-bdba-f741f3b61bb6)

JemmEx exception 06:

![image](https://github.com/SvarDOS/bugz/assets/15858666/9d4ebe7d-7a40-4226-bddb-284e509f746c)

**Comment by bttrx** [comment]
May be just coincidence, but after updating from SvarCOM 2023.1 to 2024.1 on my 'SvarDOS production system' yesterday, I can't reproduce both issues (garbage after leaving FED and 'exception 06' after leaving EDIT) anymore.

**Comment by mateuszviste** [comment]
> May be just coincidence, but after updating from SvarCOM 2023.1 to 2024.1 on my 'SvarDOS production system' yesterday, I can't reproduce both issues (garbage after leaving FED and 'exception 06' after leaving EDIT) anymore.

I confirm this seems to have been fixed as a side effect. I'm not sure at which time it got fixed exactly since I remember testing this during #38 and the issue was still presents, but clearly today it works.

This morning I retested the FREEDOS13.IMG image that you kindly prepared (very helpful, thanks!) and reproduced the issue. Then I upgraded SvarCOM to version 2024.2, rebooted, and YES - both VC 4.05 and VC 4.99.08 alpha work flawlessly now - ie. I can launch sved from within VC, quit, and VC is spawned back again without any glitching or other issues. Awesome.
